NanNew<FunctionTemplate>(foo) Cannot resolve ambiguity any more.

`NanNew<FunctionTemplate>` gets fooled by overloaded names. This used to work. https://github.com/polotek/libxmljs/pull/279

Can it be fixed somehow? The only valid signature for the first argument should be `FunctionCallback` (the signature produced by `NAN_METHOD`).

@agnat: Any ideas?

```c++
static v8::Handle<v8::Object> New(xmlNode* node);
static NAN_METHOD(New);
...
NanNew<v8::FunctionTemplate>(New);
```

```
../src/xml_comment.cc: In static member function ‘static void libxmljs::XmlComment::Initialize(v8::Handle<v8::Object>)’:
../src/xml_comment.cc:112:73: error: no matching function for call to ‘NanNew(<unresolved overloaded function type>)’
     v8::Local<v8::FunctionTemplate> t = NanNew<v8::FunctionTemplate>(New);
                                                                         ^
../src/xml_comment.cc:112:73: note: candidates are:
In file included from ../node_modules/nan/nan_new.h:165:0,
                 from ../node_modules/nan/nan.h:63,
                 from ../src/libxmljs.h:7,
                 from ../src/xml_comment.cc:5:
../node_modules/nan/nan_implementation_12_inl.h:218:1: note: template<class T> v8::Local<T> NanNew(v8::Handle<T>)
 NanNew(v8::Handle<T> h) {
 ^
../node_modules/nan/nan_implementation_12_inl.h:218:1: note:   template argument deduction/substitution failed:
../src/xml_comment.cc:112:73: note:   cannot convert ‘libxmljs::XmlComment::New’ (type ‘<unresolved overloaded function type>’) to type ‘v8::Handle<v8::FunctionTemplate>’
     v8::Local<v8::FunctionTemplate> t = NanNew<v8::FunctionTemplate>(New);
                                                                         ^
In file included from ../node_modules/nan/nan_new.h:165:0,
                 from ../node_modules/nan/nan.h:63,
                 from ../src/libxmljs.h:7,
                 from ../src/xml_comment.cc:5:
../node_modules/nan/nan_implementation_12_inl.h:224:1: note: template<class T> v8::Local<T> NanNew(const v8::Persistent<T, v8::NonCopyablePersistentTraits<T> >&)
 NanNew(v8::Persistent<T> const& p) {
 ^
../node_modules/nan/nan_implementation_12_inl.h:224:1: note:   template argument deduction/substitution failed:
../src/xml_comment.cc:112:73: note:   cannot convert ‘libxmljs::XmlComment::New’ (type ‘<unresolved overloaded function type>’) to type ‘const v8::Persistent<v8::FunctionTemplate>&’
     v8::Local<v8::FunctionTemplate> t = NanNew<v8::FunctionTemplate>(New);
                                                                         ^
In file included from ../node_modules/nan/nan.h:63:0,
                 from ../src/libxmljs.h:7,
                 from ../src/xml_comment.cc:5:
../node_modules/nan/nan_new.h:177:1: note: template<class T> typename NanIntern::Factory<T>::return_t NanNew()
 NanNew() {
 ^
../node_modules/nan/nan_new.h:177:1: note:   template argument deduction/substitution failed:
../src/xml_comment.cc:112:73: note:   candidate expects 0 arguments, 1 provided
     v8::Local<v8::FunctionTemplate> t = NanNew<v8::FunctionTemplate>(New);
                                                                         ^
In file included from ../node_modules/nan/nan.h:63:0,
                 from ../src/libxmljs.h:7,
                 from ../src/xml_comment.cc:5:
../node_modules/nan/nan_new.h:183:1: note: template<class T, class A0> typename NanIntern::Factory<T>::return_t NanNew(A0)
 NanNew(A0 arg0) {
 ^
../node_modules/nan/nan_new.h:183:1: note:   template argument deduction/substitution failed:
../src/xml_comment.cc:112:73: note:   couldn't deduce template parameter ‘A0’
     v8::Local<v8::FunctionTemplate> t = NanNew<v8::FunctionTemplate>(New);
                                                                         ^
In file included from ../node_modules/nan/nan.h:63:0,
                 from ../src/libxmljs.h:7,
                 from ../src/xml_comment.cc:5:
../node_modules/nan/nan_new.h:189:1: note: template<class T, class A0, class A1> typename NanIntern::Factory<T>::return_t NanNew(A0, A1)
 NanNew(A0 arg0, A1 arg1) {
 ^
../node_modules/nan/nan_new.h:189:1: note:   template argument deduction/substitution failed:
../src/xml_comment.cc:112:73: note:   candidate expects 2 arguments, 1 provided
     v8::Local<v8::FunctionTemplate> t = NanNew<v8::FunctionTemplate>(New);
                                                                         ^
In file included from ../node_modules/nan/nan.h:63:0,
                 from ../src/libxmljs.h:7,
                 from ../src/xml_comment.cc:5:
../node_modules/nan/nan_new.h:195:1: note: template<class T, class A0, class A1, class A2> typename NanIntern::Factory<T>::return_t NanNew(A0, A1, A2)
 NanNew(A0 arg0, A1 arg1, A2 arg2) {
 ^
../node_modules/nan/nan_new.h:195:1: note:   template argument deduction/substitution failed:
../src/xml_comment.cc:112:73: note:   candidate expects 3 arguments, 1 provided
     v8::Local<v8::FunctionTemplate> t = NanNew<v8::FunctionTemplate>(New);
                                                                         ^
In file included from ../node_modules/nan/nan.h:63:0,
                 from ../src/libxmljs.h:7,
                 from ../src/xml_comment.cc:5:
../node_modules/nan/nan_new.h:201:1: note: template<class T, class A0, class A1, class A2, class A3> typename NanIntern::Factory<T>::return_t NanNew(A0, A1, A2, A3)
 NanNew(A0 arg0, A1 arg1, A2 arg2, A3 arg3) {
 ^
../node_modules/nan/nan_new.h:201:1: note:   template argument deduction/substitution failed:
../src/xml_comment.cc:112:73: note:   candidate expects 4 arguments, 1 provided
     v8::Local<v8::FunctionTemplate> t = NanNew<v8::FunctionTemplate>(New);
                                                                         ^
```

--- a/nan_new.h
+++ b/nan_new.h
@@ -203,6 +203,14 @@ NanNew(A0 arg0, A1 arg1, A2 arg2, A3 arg3) {
   return NanIntern::Factory<T>::New(arg0, arg1, arg2, arg3);
 }
 
+template <typename T>
+typename NanIntern::Factory<T>::return_t
+NanNew( NanFunctionCallback callback
+      , v8::Handle<v8::Value> data = v8::Handle<v8::Value>()
+      , v8::Handle<v8::Signature> signature = v8::Handle<v8::Signature>()) {
+    return NanIntern::Factory<T>::New(callback, data, signature);
+}
+
 // Convenience
 
 template <typename T> inline v8::Local<T> NanNew(v8::Handle<T> h);

--- a/test/cpp/nannew.cpp
+++ b/test/cpp/nannew.cpp
@@ -328,6 +328,31 @@ NAN_METHOD(testRegression212) {
   return_NanUndefined();
 }
 
+/* Compile time regression test for https://github.com/rvagg/nan/issues/242
+ * In the presence of overloaded functions NaN should be able to pick the one
+ * matching NanFunctionCallback. 
+ */
+void overloaded() {}
+NAN_METHOD(overloaded) {
+    overloaded();  // not unused
+    return_NanUndefined();
+}
+
+NAN_METHOD(testRegression242) {
+  NanScope();
+  NanTap t(args[0]);
+
+  // This line needs to *compile*. Not much to test at runtime.
+  Local<FunctionTemplate> ft = NanNew<FunctionTemplate>(overloaded);
+  (void)ft;  // not unused
+
+  t.plan(1);
+
+  t.ok(true, "compile-time regression test #242");
+
+  return_NanUndefined();
+}
+
 
 //==============================================================================
 // JavaScript Tests
@@ -386,6 +411,7 @@ void Init(Handle<Object> exports) {
   NAN_EXPORT(exports, testPersistents);
 
   NAN_EXPORT(exports, testRegression212);
+  NAN_EXPORT(exports, testRegression242);
 
   NAN_EXPORT(exports, newIntegerWithValue);
   NAN_EXPORT(exports, newNumberWithValue);

